drivers/counter/counter_nrfx_timer.c

```diff
@@ -0,0 +1,282 @@
+		return -EINVAL;
+	}
+
+	if (nrfx_config->alarm_cfgs[alarm_cfg->channel_id]) {
+		return -EBUSY;
+	}
+
+	cc_val = alarm_cfg->ticks + (alarm_cfg->absolute ?
+				0 : nrfx_timer_capture(timer, COUNTER_READ_CC));
+	nrfx_config->alarm_cfgs[alarm_cfg->channel_id] = alarm_cfg;
+
+	nrfx_timer_compare(timer, ID_TO_CC(alarm_cfg->channel_id),
+			   cc_val, true);
+
+	return 0;
+}
+
+static void _disable(struct device *dev, u8_t id)
+{
+	const struct counter_nrfx_config *config = get_nrfx_config(dev);
+
+	nrfx_timer_compare_int_disable(&config->timer, ID_TO_CC(id));
+	config->alarm_cfgs[id] = NULL;
+}
+
+static int counter_nrfx_disable_alarm(struct device *dev,
+				      const struct counter_alarm_cfg *alarm_cfg)
+{
+	_disable(dev, alarm_cfg->channel_id);
+
+	return 0;
+}
+
+static int counter_nrfx_set_wrap_alarm(struct device *dev,
+				       u32_t ticks,
+				       counter_wrap_callback_t callback,
+				       void *user_data)
+{
+	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
+	const nrfx_timer_t *timer = &nrfx_config->timer;
+	struct counter_nrfx_data *data = get_dev_data(dev);
+
+	for (int i = 0; i < counter_get_num_of_channels(dev); i++) {
+		/* Overflow can be changed only when all alarms are
+		 * disables.
+		 */
+		if (nrfx_config->alarm_cfgs[i]) {
+			return -EBUSY;
+		}
+	}
+
+	nrfx_timer_compare_int_disable(timer, WRAP_CH);
+	nrfx_timer_clear(timer);
+
+	data->wrap_cb = callback;
+	data->wrap_user_data = user_data;
+	nrfx_timer_extended_compare(timer, WRAP_CH,
+				    ticks, COUNTER_OVERFLOW_SHORT,
+				    callback ? true : false);
+
+	return 0;
+}
+
+static u32_t counter_nrfx_get_pending_int(struct device *dev)
+{
+	return 0;
+}
+
+static void alarm_event_handler(struct device *dev, u32_t id)
+{
+	const nrfx_timer_t *timer = &get_nrfx_config(dev)->timer;
+	const struct counter_alarm_cfg *alarm_cfg =
+				get_nrfx_config(dev)->alarm_cfgs[id];
+	u32_t cc_val;
+
+	if (!alarm_cfg->handler) {
+		return;
+	}
+
+	cc_val = nrfx_timer_capture_get(timer, ID_TO_CC(id));
+	_disable(dev, id);
+	alarm_cfg->handler(dev, alarm_cfg, cc_val);
+}
+
+static void event_handler(nrf_timer_event_t event_type, void *p_context)
+{
+	struct device *dev = p_context;
+	struct counter_nrfx_data *dev_data = get_dev_data(dev);
+
+	if (event_type == COUNTER_WRAP_INT) {
+		if (dev_data->wrap_cb) {
+			dev_data->wrap_cb(dev, dev_data->wrap_user_data);
+		}
+	} else if (event_type > NRF_TIMER_EVENT_COMPARE1) {
+		alarm_event_handler(dev, COUNTER_EVENT_TO_ID(event_type));
+
+	}
+}
+
+static int init_timer(struct device *dev, const nrfx_timer_config_t *config)
+{
+	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
+	const nrfx_timer_t *timer = &nrfx_config->timer;
+
+	nrfx_err_t result = nrfx_timer_init(timer, config, event_handler);
+
+	if (result != NRFX_SUCCESS) {
+		LOG_INST_ERR(nrfx_config->log, "Failed to initialize device.");
+		return -EBUSY;
+	}
+
+	nrfx_timer_compare(timer, WRAP_CH, UINT32_MAX, false);
```

Comment:
This way we don't use the full range of the counter (one tick is actually wasted, it's not that much but still). Maybe it would be good to describe in the API that max_wrap equal to 0 means full 32-bit range (i.e. the counter wraps to 0 after increasing from UINT32_MAX, or wraps to UINT32_MAX after decreasing from 0 when counting down)?